src/Mod/Sketcher/Gui/DrawSketchController.h

```diff
@@ -452,6 +453,18 @@ class DrawSketchController
         }
     }
 
+    bool shouldDrawPositionAtCursor() const
+    {
+        return !(ovpVisibilityManager.onViewParameterVisibility
+                 == OnViewParameterVisibilityManager::OnViewParameterVisibility::ShowAll);
+    }
+
+    bool shouldDrawDimensionsAtCursor() const
+    {
+        return (ovpVisibilityManager.onViewParameterVisibility
+                == OnViewParameterVisibilityManager::OnViewParameterVisibility::Hidden);
+    }
+

```

Comment:
This here is an option. It is not my preferred choice.

The rationale to have this here, is that the "logic" for the control of what to do  (show or not show) is with the controller.

An implication of this choice, is on the caller (DrawSketchControllableHandler). When the functionality is triggered from the Handler, by checking the controller (the present model here), a new interface is being imposed on any future controller that is written to work with DrawSketchControllableHandler. That includes controllers that may not have any constraint on whether to show or not a cursor.

I do not think it is inherently wrong. I have an alternative that I will comment on below.

Reply:
Ok.
But ultimately will there be any other controllers ? I think splitting DSH in many classes and subclasses for the sake of potential future handlers that could have different need is not really useful and counter productive. And I think there is also a coding holy principle about this even though I forgot the name, something like 'handle what is needed now and if the need appears later it will be adapted at that time'.
Besides sketcher is mostly mature. And the few tools that were missing are being added now and don't need a special controller.
So for me DSH, DSDH, DSCH should all be squashed. The architecture you build is probably very conceptually sound, but that is actually complex in the amount of files / subclasses / template etc and that only you understand clearly. I worked on it for more than one year and I'm often confused and lost between the files.

Here I'm a bit out of topic, but actually it is on topic. Because look with the complexity castle you build, my PR is not implementing exactly at the right place and so on. What is the likelyhood that a newcomer will be able to get in FC code and do what I've done correctly ? None. Even other fc devs would get it wrong I think.
So in the end this complexity is against the KISS principle (Keep it simple, stupid). Which in the end makes the code less accessible.

Reply:
Not saying that your arguments are not sound here considering the current architecture. Just saying that the architecture is so complex that if I get it wrong after all the time I've been in it, this might be an indication that complexity is too high. And that we could probably reduce it.

Reply:
**Would there by any other controllers?** 

It is a valid question. 

You negative answer lies on the premise that no new tools will be added as the sketcher is mature. But you have added new DSHs haven't you? When the premise is that a framework does not need to be extended, there is no need to design it allow extension. Yet, what almost systematically ends up happening, is that software grows and needs be extended. When architecture does not support growth, the problem is that developers start hacking around to solve their immediate needs, which opens the can of worms.

An answer has to take into account how much sketcherdefaultwidget will be able to grow without becoming a Frankenstein (and I honestly think it is close). That class holds a bunch of generally used controls, which is not certainly all the controls possible. I would say the class is useful for writing many DSHs, but it does not scale to cover the widget needs of ANY DSH.

In other words, I do not think it is unthinkable that some DSHs would not be feasible without allowing to provide a custom widget.

Now the principle you were looking for is probably YAGNI (You ain't gonna need it). It is a possible line of argumentation. Then one should think the cost of splitting the functionality into classes, with separate responsibilities now, and the cost of doing it when the need arises. Given the complexity of the framework, I think it totally made sense to do it now.

**Framework complexity** 

Now, you are proposing to squash DrawSketchHandler, DrawSketchDefaultHandler and DrawSketchControllableHandler. Your premise is that putting all into one class, as opposed to separate them by responsibilities with an eye in extensibility, makes the code simpler. 

Let me answer you with one question: Why not write every handler from scratch? Just create a class DrawSketchHandlerLine and write all the code there. No code reuse. Rinse and repeat for DrawSketchHandlerCircle and the rest. Hopefully, you will say, hey no, I need code reuse. Fine.

The answer of why these classes exist separately is similar:
- DrawSketchHandler is the old framework, which allows to write any type of handler. Not only Geometry creation handlers.
- DrawSketchDefaultHandler is a new type of handler, which has a very specific structure useful to add geometry (along with constraints and autconstraints) and offering many advanced solver capabilities. This class is intended to be an improved DrawSketchHandler specialised for this type of handler. A person wanting to add a handler for geometry creation today, and not wanting OVPs or widgets, should derived from this one instead.
- DrawSketchControllableHandler was created because your specific need to have double input (mouse+keyboard), first for the parameters in the widget and then for the OVPs. This class is exactly the same as the previous one, but allows to exchange changes between handler and controller (for OVPs and Widget). The controller can be anything that meets its interface.
- DrawSketchController is the most basic controller supporting the interface of DrawSketchControllableHandler and provides the OVPs. A developer can derive from it if it wants a controller for OVPs, but no default widget, or a custom widget.
- DrawSketchDefaultWidgetController is a controller that in addition to OVPs uses the default widget for input.

You have all these classes so that the framework is flexible, and each of the classes is able to concentrate in one task. Some of this classes are templates to enable quite a lot of automation, while minimising typing.

So, imagine we do as you say, and make them all into one class. What would you be winning? Yes, you will have less amount of classes. But still you need to handle the default widget, and the OVPs and the dual input. So at the end, you do need all that code. You do not have a simpler framework. You do have a less flexible framework. You still need to deal with the same complexity, because no matter how you distribute the code, code needs to be maintained (framework) and DSH code needs to be provided.

In fact, what you should think is how much code are you saving everytime you write a new tool. Yes, files a bigger with the dual input... but do you think they would be smaller by keeping all the framework in a single class. No, they would not.

All the code of specialised functions is **because of the dual input and the widget**.

I went into a lot of thinking an trouble to handle this complexity of the dual input in a uniform way, and which as much code reuse as possible.

**Dealing with the complexity of the framework** 

When the functionality demanded to a framework is complex, no matter how much architecture you factor in, the framework will be complex. Paraphrasing Einstein roughly, **things should be as simple as possible, but not simpler**.

You claim you have problems to change the framework. Well, you have changed the framework, so you managed. You managed even if you do not appreciate its architecture... should not that speak for it being a good framework? After all, a complex framework that is fixable by a developer that is not the one who wrote it, should speak for its virtues, rather than it shortcomings.

In any case, the objective of a framework is not that is simpler than need be internally, but rather that it abstracts a client of the framework from the complexity it was written to handle. 

Hopefully, we will have over time developers in FreeCAD who appreciate and understand the need for an architecture. If we fail there, then it is guaranteed that FreeCAD will turn into an unmaintainable beast. This has happened before in many projects, including enterprise projects employing good programmers. That motivated people to write books about software architecture.

So my answer here is: We hope to provide a framework new developers can directly rely on for a vast variety of DSHs, and concentrate their efforts on the handlers themselves. If the framework does not adapt to a particular need, this requires a more experienced developer, or one who can understand the need for an architecture.

**Is it difficult to write DSHs using the framework?** 

This is a trap question. Difficult compared to what?. 

It is certainly more difficult to write a new DSH having OVPs and a widget than it was to write one of the legacy DSHs which have none of those facilities. However, to compare pears to pears, it is not more difficult to write a new handler without those facilities using DrawSketchDefaultHandler, and it provides a lot more facilities (advanced solver functionality, automatic drawing based on geometries rather than generating a sampled curve, ...), than using DrawSketchHandler (specially where there are examples).

Now, start deriving from DrawSketchDefaultHandler (or DrawSketchHandler if you so wish) and without using the controllable handler or the written controllers, try to write a class handling OVPs and widget. Is that much more complex? The answer can only be that it is indeed much more complex, because part of the complexity (a substantial amount of it) is in those classes. The result would also be more code without code reuse (less concise). More error prone. More code to maintain. Duplicated code over handlers...

To reply to your last post. I do not think this architecture is more complex that should be for the requirements imposed upon it. The words are not mine, but are very wise words (it may well be because they are not mine): **things should be as simple as possible, but not simpler.**

N.B.: BTW, I did not tell you your solution was abhorrent. I only told you why I would prefer a different one. I took the time to explain and even code alternatives, not because I have too much free time. But because I prefer you to understand the choices, so that your code gets better.

Reply:
> Writing a DSH with the current architecture is not a problem. It's more debugging/changing the architecture that is harder. Like when offset was crashing, it felt like a maze.

A good framework makes the life of the client code the easiest possible. A good framework needs to manage its complexity in the most simple way, but cannot be simpler than the minimum required by the requirements imposed upon it by design.

Yes, I agree that debugging the internals of a good framework which manages a complex task is not easy. The motivation to debug and fix any bug is that, when it is fixed, it will work, and using it will be very satisfactory. The extra motivation is that by leveraging complexity with the framework, you fix it once for all the tools using the framework.

> Ok so maybe not the base DSH, but why then DrawSketchControllableHandler and DrawSketchDefaultHandler. If a tool don't need widget/ovp, why couldn't it just use DrawSketchControllableHandler and declare controller null. That would remove one layer.

DrawSketchControllableHandler defines an interface of the controller, which is a template type. To implement what you suggest, you would need to create a controller that has the interface required by that class (and maintain it, so every change to the interface, needs a change to that fake controller), and make the fake controller do nothing it. Having code doing nothing does not appear to be removing complexity. At the same time, you had to add a new class (fake controller) to remove another class (merging two into one). So not even the number of classes decreases (which in any case is a faux measure of complexity).

There are several reasons why this is not a good idea. But perhaps it is best to try to understand why you think it is a good idea. Does removing one class in the hierarchy make dealing with the complexity any simpler? Why? It is the same amount of code and it is intermixed in a single class, as opposed to organised into areas of responsibility. The only thing that the controllable handler does is deal with the signaling necessary to interface with the controller. Anything else you look into daddy class... or grandpa class.

A first reason why this is a bad idea, is that in a hierarchy of classes, each derived class goes developing an extra aspect, and holds the responsibility for managing this aspect. This should help in the sense that if you want to fix a given functionality, you go to the class holding the responsibility. If you need to review/fix/extend a given functionality and you have merged ten classes into one, you will end up reading ten times more code. This is a maintainability killer.

A second reason why I think it is a very bad idea to merge these two classes, is because if a developer wants to make a simple DSH with the new architecture (without OVPs and widgets), and looks into this merged parent, this person will see much more code than actually needs see... and for nothing. Moreover, this person will be forced to create a dummy controller and assign it... which is, at best, odd. This class shows a very bad intent/interface to the user. The objective should be to make the framework the easiest possible to use by the user, while encapsulating complexity. Forcing the user to add a dummy controller is at odds with this goal.

A third reason, is that by doing so you remove flexibility and extensibility from the framework. We are now in a situation where we are extending DSDefaultHandler to have OVPs and widgets. If anybody would have a different need in the future, with a single class, this person would be forced to carry on with all that controller specific code, even if this person does not want it. While developing this need, this person will have to debug too, and it is unlikely he or she would appreciate seeing framework code that makes no sense whatsoever. This is making the person deal with more complexity than really needed.

There are probably other reasons, but I have been too long awake...

Reply:
I understand thanks. I think I have just been frustrated by the complexity when debugging especially considering the number of time I've done so. It made me feel like there is a hundred OnReset function each overridding each other. But yes squashing classes will probably not help, its wishful thinking.
As you said, its making the framework which is understandably hard. While writing dsh is indeed easy enough.